Core: Switch tests to Junit5 in rest, hadoop pakages

skytin1004 opened the pull request:
This pull request is related to the issue "Move JUnit4 tests to JUnit5 https://github.com/apache/iceberg/issues/7160".
I switched the files in [rest, hadoop] pakages to JUnit5.

All files in Core: rest package have completed the switch to JUnit5.
I have switched some `Hadoop packages` to JUnit5 and switched all `Hadoop packages` from `Assert` to `AssertJ`.

In Core: Hadoop package, when switching `"HadoopTableTestBase.java"`, to JUnit5, an error occurred in`"testListNamespace"` of`"TestHadoopCatalog.java"` where the namespace of the root directory of the` temp` folder could not be looked up during testing.

So, I switched only the files that are not related to `"HadoopTableTestBase.java"` in the Hadoop package to JUnit5, kept the files related to `"HadoopTableTestBase.java"` in JUnit4, and switched all the parts that are written in `Assert` to `AssertJ`.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestTableSerialization.java:
```diff
@@ -41,7 +41,7 @@
 import org.apache.iceberg.types.Types;
-import org.junit.Assert;
+import org.assertj.core.api.Assertions;
 import org.junit.Test;
```
needs to be switched to Junit5

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestStaticTable.java:
```diff
@@ -24,7 +24,6 @@
 import org.apache.iceberg.relocated.com.google.common.collect.Maps;
 import org.assertj.core.api.Assertions;
-import org.junit.Assert;
 import org.junit.Test;
```
needs to be switched to Junit5

skytin1004 replied:
I have switched to JUnit5.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestStaticTable.java:
```diff
@@ -88,13 +87,13 @@ public void testHasSameProperties() {
+    Assertions.assertThat(table.currentSnapshot().snapshotId())
+        .as("Same snapshot?")
+        .isEqualTo(staticTable.currentSnapshot().snapshotId());
+    Assertions.assertThat(Maps.difference(table.properties(), staticTable.properties()).areEqual())
```
can we express that with an AssertJ assertion rather than using isTrue?

skytin1004 replied:
I've modified that code based on what you suggested.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -55,20 +54,19 @@ public class TestHadoopTables {
           required(2, "data", Types.StringType.get()));
 
-  @Rule public TemporaryFolder temp = new TemporaryFolder();
   private File tableDir = null;
```
I think we can just add a `@TempDir` annotation to `tableDir` and remove `@TempDir Path temp` alltogether?

skytin1004 replied:
I have switched to @TempDir tableDir

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -57,7 +58,6 @@
 import org.apache.iceberg.util.Tasks;
 import org.assertj.core.api.Assertions;
-import org.junit.Assert;
 import org.junit.Test;
```
needs to be switched to Junit5

skytin1004 replied:
I switched to JUnit5.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -55,20 +54,19 @@ public class TestHadoopTables {
-  public void setupTableLocation() throws Exception {
-    tableDir = temp.newFolder();
+  @BeforeEach
+  public void setupTableLocation(@TempDir Path temp) throws Exception {
```
the entire `setupTableLocation()` can be removed I believe

skytin1004 replied:
Thank you for your review. I have now fixed the code based on your review and switched the file to JUnit5. I will review all your reviews and post a commit.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -92,15 +90,15 @@ public void testDropTableWithPurge() throws IOException {
   @Test
-  public void testDropTableWithoutPurge() throws IOException {
-    File dataDir = temp.newFolder();
+  public void testDropTableWithoutPurge(@TempDir Path temp) throws IOException {
```
why do we need the `@TempDir Path temp` parameter here? We should be able to use `tableDir`

skytin1004 replied:
I replaced them all with tableDir.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestCachingCatalog.java:
```diff
@@ -291,9 +292,10 @@ public void testCacheExpirationEagerlyRemovesMetadataTables() throws IOException
-                Assert.assertFalse(
-                    "When a data table expires, its metadata tables should expire regardless of age",
-                    catalog.cache().asMap().containsKey(metadataTable)));
+                Assertions.assertThat(catalog.cache().asMap().containsKey(metadataTable))
```
I modified the code based on the method you suggested. `.doesNotContainKeys(metadataTable)`

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/HadoopTableTestBase.java:
```diff
@@ -113,17 +113,17 @@ public class HadoopTableTestBase {
           .build();
 
-  @Rule public TemporaryFolder temp = new TemporaryFolder();
+  @TempDir Path temp;
```
```suggestion
  @TempDir File tableDir;
```
and then we can remove all `this.tableDir = temp.toFile()` calls

skytin1004 replied:
Thank you for the suggestion. I've modified the code based on your suggestions.

skytin1004 commented:
Thank you so much for your very detailed review @nastra .
Based on your review, I was able to change the Hadoop package to JUnit5, incorporating all the things you reviewed.
However, there is an issue. When I run the tests in TestHadoopCatalog, I get the following results 
The test "testListNamespace" fails.
I investigated the problem through debug and found that  the namespaces "db", "db2" in the root directory are not being looked up properly, since switching to JUnit5 In that method.
So I would like to get some Solutions from you. Thank you again for the detailed review :)

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -51,7 +51,6 @@
 import org.apache.iceberg.transforms.Transforms;
 import org.assertj.core.api.Assertions;
-import org.junit.Assert;
 import org.junit.Test;
```
I've modified that code based on your review, thank you.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestCatalogUtilDropTable.java:
```diff
@@ -124,8 +131,12 @@ public void shouldNotDropDataFilesIfGcNotEnabled() {
+        .isEqualTo(2);
+    Assertions.assertThat(metadataLocations.size())
+        .as("should have 4 metadata locations")
+        .isEqualTo(4);
```
I've modified that code based on your review.

nastra reviewed (commented):
You'll need these changes to make the failing test work
```
+  @TempDir File tempDir;
   @TempDir File tableDir;
 
   String tableLocation = null;
@@ -121,7 +122,6 @@ public class HadoopTableTestBase {
 
   @BeforeEach
   public void setupTable() throws Exception {
-
     this.tableLocation = tableDir.toURI().toString();
     this.metadataDir = new File(tableDir, "metadata");
     this.versionHintFile = new File(metadataDir, "version-hint.text");
@@ -193,7 +193,7 @@ public class HadoopTableTestBase {
         "hadoop",
         ImmutableMap.<String, String>builder()
             .putAll(catalogProperties)
-            .put(CatalogProperties.WAREHOUSE_LOCATION, tableDir.getAbsolutePath())
+            .put(CatalogProperties.WAREHOUSE_LOCATION, tempDir.getAbsolutePath())
```

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -105,23 +104,25 @@ public void testReplaceTxnBuilder() throws Exception {
 
-    Assert.assertEquals("value1", table.properties().get("key1"));
-    Assert.assertEquals("value2", table.properties().get("key2"));
+    Assertions.assertThat(table.properties().get("key1")).isEqualTo("value1");
```
I modified the code based on the method you suggested.

skytin1004 commented on core/src/test/java/org/apache/iceberg/rest/RequestResponseTestBase.java:
```diff
@@ -76,11 +75,13 @@ public void testHasOnlyKnownFields() {
 
       for (String field : ((Iterable<? extends String>) node::fieldNames)) {
-        Assert.assertTrue("Should not have field: " + field, fieldsFromSpec.contains(field));
+        Assertions.assertThat(fieldsFromSpec.contains(field))
```
I modified the code based on the method you presented.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -55,104 +53,103 @@ public class TestHadoopTables {
-  public void setupTableLocation() throws Exception {
-    tableDir = temp.newFolder();
-  }
+  @TempDir private Path tableDir;
```
Based on your review, I've changed the code and deleted toUri() changes.

skytin1004 commented:
Once again, thank you so much for your detailed review @nastra  . I made some code changes based on your suggestions and the test issue was resolved.
And I modified the code based on your review, using 
`@TempDir private File tableDir;
@TempDir private Path dataDir;` 
to eliminate unnecessary `toUri()`changes.

And I modified the code based on the method `.containsEntry(...).containsEntry(..)` you suggested to

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -69,10 +68,10 @@ public void testCreateTableBuilder() throws Exception {
-    Assert.assertEquals("value2", table.properties().get("key2"));
+    Assertions.assertThat(table.schema().toString()).isEqualTo(TABLE_SCHEMA.toString());
+    Assertions.assertThat(table.spec().fields()).hasSize(1);
+    Assertions.assertThat(table.properties().get("key1")).isEqualTo("value1");
```
those should all be `.containsEntry` (please also double-check other places that are being modified as part of this PR so that we use the same mechanism everywhere)

skytin1004 replied:
I double-checked and changed everything that could be changed to .containsEntry.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -404,14 +444,14 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
-    File dir = temp.newFolder();
-    dir.delete();
+    File dir = tableDir;
+    FileUtils.deleteDirectory(dir);
```
`Table already exists at location: file:/tmp/junit5635809745002339218/`

nastra replied:
then it's probably better to inject a separate temp dir for this particular test using `testConcurrentFastAppends(@TempDir File dir)` as a method parameter

skytin1004 replied:
Thank you for the suggestion, I've modified the code based on your suggestion.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -267,13 +273,13 @@ public void testListTables() throws Exception {
-    Assert.assertEquals(2, tblSet.size());
-    Assert.assertTrue(tblSet.contains("tbl1"));
-    Assert.assertTrue(tblSet.contains("tbl2"));
+    Assertions.assertThat(tblSet).hasSize(2);
```
I went through all of my code and shortened wherever I could.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/HadoopFileIOTest.java:
```diff
@@ -111,16 +110,16 @@ public void testDeletePrefix() {
-                  () -> hadoopFileIO.listPrefix(scalePath.toUri().toString()).iterator());
+              Assertions.assertThatThrownBy(
+                      () -> hadoopFileIO.listPrefix(scalePath.toUri().toString()).iterator())
+                  .isInstanceOf(UncheckedIOException.class);
```
can we add a `.hasMessage(..)` check here and below?

skytin1004 replied:
I'll add the `.hasmessage` .

skytin1004 replied:
Based on your review, I added the hasMessage in `Assertions.assertThatThrownBy` .

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -105,23 +104,25 @@ public void testReplaceTxnBuilder() throws Exception {
-    Assert.assertEquals("value1", table.properties().get("key1"));
-    Assert.assertEquals("value2", table.properties().get("key2"));
+    Assertions.assertThat(table.properties()).containsEntry("key1", "value1");
+    Assertions.assertThat(table.properties()).containsEntry("key2", "value2");
```
can be chained together

skytin1004 replied:
Based on your review, I've modified the code to be more concise.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -67,55 +64,59 @@ public class TestHadoopCommits extends HadoopTableTestBase {
-
+    Assertions.assertThat(tasks).as("Should not create any scan tasks").isEmpty();
+    Assertions.assertThat(tableDir).as("Table location should exist").exists();
+    Assertions.assertThat(metadataDir.exists() && metadataDir.isDirectory())
```
Assertions.assertThat(metadataDir).exists().isDirectory()

skytin1004 replied:
I've modified the code based on your review in this PR

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -67,55 +64,59 @@ public class TestHadoopCommits extends HadoopTableTestBase {
+    Assertions.assertThat(metadataDir.exists() && metadataDir.isDirectory())
+        .as("Should create metadata folder")
+        .isTrue();
+    Assertions.assertThat(version(1).exists() && version(1).isFile())
```
```suggestion
    Assertions.assertThat(version(1)).exists().isFile();
```

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -67,55 +64,59 @@ public class TestHadoopCommits extends HadoopTableTestBase {
+    Assertions.assertThat(version(1).exists() && version(1).isFile())
+        .as("Should create v1 metadata")
+        .isTrue();
+    Assertions.assertThat(version(2).exists())
```
```suggestion
    Assertions.assertThat(version(2)).exists()
```

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -57,8 +58,7 @@
 import org.assertj.core.api.Assertions;
-import org.junit.Assert;
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
```
Based on your review, I've changed all the remaining JUnit4 code in this PR to JUnit5.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -55,104 +53,101 @@ public class TestHadoopTables {
 
   @Test
   public void testTableExists() {
-    Assert.assertFalse(TABLES.exists(tableDir.toURI().toString()));
```
Based on your review, I've restored the `toURI()` calls to its original state. It looks like I would misunderstood.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -83,36 +76,32 @@ public void testDropTable() {
-    File dataDir = temp.newFolder();
 
-    createDummyTable(tableDir, dataDir);
+    createDummyTable(tableDir, dataDir.toFile());
```
I've changed dataDir to File instead of Path

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -92,27 +83,23 @@ public void testDropTableWithPurge() throws IOException {
-    Assert.assertFalse(tableDir.exists());
-
-    Assert.assertFalse(TABLES.dropTable(tableDir.toURI().toString()));
+    Assertions.assertThat(Files.list(dataDir.toPath()).count()).isEqualTo(0);
```
given that we switched back dataDir to a File, we don't need to call `Files.list(..)` here right?

skytin1004 replied:
Your review is correct. I'll modify the code now.

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopTables.java:
```diff
@@ -92,27 +83,23 @@ public void testDropTableWithPurge() throws IOException {
-    Assert.assertFalse(tableDir.exists());
-
-    Assert.assertFalse(TABLES.dropTable(tableDir.toURI().toString()));
+    Assertions.assertThat(Files.list(dataDir.toPath()).count()).isEqualTo(1);
```
same here

skytin1004 replied:
I've modified `Files.list(dataDir.toPath()).count()` to

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -69,10 +68,10 @@ public void testCreateTableBuilder() throws Exception {
+    Assertions.assertThat(table.schema().toString()).isEqualTo(TABLE_SCHEMA.toString());
+    Assertions.assertThat(table.spec().fields()).hasSize(1);
+    Assertions.assertThat(table.properties()).containsEntry("key1", "value1");
+    Assertions.assertThat(table.properties()).containsEntry("key2", "value2");
```
Based on your review, I've modified the code to be more concise.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -530,13 +533,14 @@ public void testTableName() throws Exception {
 
     Table table = catalog.loadTable(tableIdent);
-    Assert.assertEquals("Name must match", "hadoop.db.ns1.ns2.tbl", table.name());
+    Assertions.assertThat(table.name()).as("Name must match").isEqualTo("hadoop.db.ns1.ns2.tbl");
```
Based on your review, I've removed unnecessary descriptions.

skytin1004 commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:
```diff
@@ -267,13 +273,11 @@ public void testListTables() throws Exception {
     List<TableIdentifier> tbls2 = catalog.listTables(Namespace.of("db", "ns1"));
-    Assert.assertEquals("table identifiers", 1, tbls2.size());
-    Assert.assertEquals("table name", "tbl3", tbls2.get(0).name());
+    Assertions.assertThat(tbls2).as("table identifiers").hasSize(1);
```
Based on your review, I've removed unnecessary descriptions.

nastra reviewed (approved):
LGTM, thanks @skytin1004

skytin1004 commented:
Thank you @nastra, I've fixed the code based on your review.
There was a lot that needed to be fixed in my PR this time around. 
I'm glad to be able to fix the problem with PR based on your detailed review.
